Add repo audit

asmith-mdsol opened the pull request:
This change introduces the `repo audit` command, which identifies all commits in the named repo that are not covered by a merged pull request. Current limitations include:
- Command must be run within a local copy of the repo (for fast log access). Need command line options to specify working directory and/or clone.
- Looks at entire repo history. Need a command line option to specify an epoch for logging.
- Assumes all pull requests are important, regardless of target branch. Need a command line option to filter to specific target branch but seems usable right now without this.
- Crashes if a pull request contains a head commit no longer present in the repo history. Still investigating how this can come about (suspect history rewriting).

Despite these limitations, I think this change is worth having as it's a good base to build the suggested options onto and it has already turned up some interesting results during testing!

@byoung Interested?

asmith-mdsol committed:
Add gem dependency git ~> 1.2

asmith-mdsol committed:
Enable auto traversal of pages in OctoKit client

This is necessary to ensure that, for example, all pull requests are
retrieved and not just the first page of 30.

asmith-mdsol committed:
Add repo audit command

asmith-mdsol committed:
Update README and version

asmith-mdsol committed:
Exclude unmerged pull requests from repo audit

Include cosmetic refactoring.

asmith-mdsol committed:
Ensure git client returns *all* commits in logs

glow-mdsol commented:
I think there are scopes for having different levels of auditing; ie 
cleric repo audit commits
cleric repo audit pull-requests [check all pull requests have comments, or are not created and merged by the same person]

Much of the plumbing is exactly the same, it's just taking a different aspect.

Could we have this as more specific (ie we're auditing the commits) from the command line?

Just a thought.

asmith-mdsol commented:
@glow-mdsol I definitely see this being broken out it aspects as you suggest in the future. Just focusing on the simplest thing that works right now.

asmith-mdsol commented:
By the way, if @glow-mdsol, @isaacwong, @chad-medi or anyone else would like to help land this, I'd appreciate it! I should probably add the following as I know Ben likes to see it:

```
All examples were filtered out; ignoring {:focus=>true}
..........................................................GitHub login: GitHub password: .....................................

Finished in 3.49 seconds
95 examples, 0 failures
```

ghost commented:
Not sure about your specs.  Too many "should_receive".  Which IMHO is evil.  I wish rspec would remove it.

asmith-mdsol commented on lib/cleric/console_announcer.rb:
```diff
@@ -10,10 +10,19 @@ def chatroom_added_to_repo(repo, chatroom)
     end
 
+    def commits_without_pull_requests(repo, commits)
+      write_warning("Repo \"#{repo}\" has the following commits not covered by pull requests:\n" +
```
If I come back past here I'll take a look. My only issue with HEREDOCs is dealing with the leading whitespace; any suggestions welcome!

asmith-mdsol commented on lib/cleric/console_announcer.rb:
```diff
@@ -10,10 +10,19 @@ def chatroom_added_to_repo(repo, chatroom)
+    end
+
     def repo_added_to_team(repo, team)
       write_success("Repo \"#{repo}\" added to team \"#{team}\"")
```
Yeah, I always did this until just recently. Guess I felt like a change. I have to admit standing back this is pretty ugly right now.

asmith-mdsol commented on lib/cleric/repo_auditor.rb:
```diff
@@ -0,0 +1,42 @@
+require 'set'
+
+module Cleric
+  class RepoAuditor
```
Will do!

asmith-mdsol commented on lib/cleric/repo_auditor.rb:
```diff
@@ -0,0 +1,42 @@
+      commits_with_pr = Set.new
+
+      ranges = repo_agent.repo_pull_request_ranges(name)
+      ranges.each do |range|
```
I have an aversion to `inject` but I guess I should get over it. I feel I should be able to do everything with `map` and if I can't I devolve to `each`!

asmith-mdsol commented on spec/cleric/cli_spec.rb:
```diff
@@ -105,6 +108,23 @@ def stub_options_for(obj, options)
+        after(:each) { repo.audit('my/repo') }
+
+        it 'creates a console announcer' do
+          ConsoleAnnouncer.should_receive(:new).with($stdout)
```
Take your point on `after :each`; I think that's definitely in the category of "easy to write, hard to read".

asmith-mdsol commented on spec/cleric/repo_auditor_spec.rb:
```diff
@@ -0,0 +1,55 @@
+        agent.should_receive(:repo_pull_request_ranges).with('my/repo')
+      end
+      it 'opens a client for the local repo' do
+        Git.should_receive(:open).with('.')
```
I'm torn.

From one perspective I think it's perfectly fine because this class' job is to talk to that library, so that's what my unit test should focus on. If I look towards testing outcomes in git, I'm going to have to set that all up externally, and it feels like I'm straying into integration testing territory.

On the other hand, I can see where you're coming from, that my tests are brittle because of this.

I don't know which way to go (and part of this project was to give me an opportunity to practice TDD!) so I'm going to leave it as-is for now.

asmith-mdsol commented:
Will address review comments from Chad but it may take a day or two for me to find the time.

asmith-mdsol committed:
Refactor string quoting

asmith-mdsol committed:
Add missing class documentation

asmith-mdsol committed:
Refactor RepoAuditor#audit_repo based on review

I must not fear `inject`. Fear is the mind-killer. Fear is the
little-death that brings total obliteration. I will face my fear of
`inject`. I will permit it to pass over me and through me. And when it
has gone past I will turn the inner eye to see its path. Where the fear
has gone there will be nothing... Only `inject` will remain.

asmith-mdsol commented:
@chad-medi Ready for landing again.

With respect, I'm leaving the specs alone as this was an intentional stylistic choice. I'm following the "mockist" style as described by Martin Fowler here:

http://martinfowler.com/articles/mocksArentStubs.html#ClassicalAndMockistTesting

I'm using this codebase to try out that style.

The ruby-git gem uses a cunningly inlined git repo for its tests. I'm happy to tightly binding my tests to a library that I trust (and am unlikely to swap out) in exchange for bypassing all that hassle of testing directly against git.
